sdk/communication/AzureCommunicationChat/Source/ChatClient.swift

```diff
@@ -99,6 +99,52 @@ public class ChatClient {
         )
 
         self.service = client.chat
+        self.signalingClient = createSignalingClient(credential: credential)
+    }
+
+    private func createSignalingClient(credential: CommunicationTokenCredential)
+        -> CommunicationSignalingClient? {
+        var signalingClient: CommunicationSignalingClient?
+        // Retrieve the access token
+        credential.token { accessToken, error in
+            do {
+                guard let token = accessToken?.token else {
+                    throw AzureError.client("Failed to get token from credential.", error)
+                }
+
+                let tokenProvider = CommunicationSkypeTokenProvider(
+                    token: token,
+                    credential: self.credential,
+                    tokenRefreshHandler: { stopSignalingClient, error in
+                        // Unable to refresh the token, stop the connection
+                        if stopSignalingClient {
+                            self.signalingClient?.stop()
+                            self.signalingClientStarted = false
+                            self.options
+                                .signalingErrorHandler?(
+                                    .failedToRefreshToken(
+                                        "Unable to get valid token for realtime-notifications, stopping notifications."
+                                    )
+                                )
+                            return
+                        }
+
+                        // Token is invalid, attempting to refresh token
+                        self.options.logger.error("Failed to get valid token. \(error ?? "")")
+                        self.options.logger.warning("Attempting to refresh token for realtime-notifications.")
+                    }
+                )
+
+                // Initialize the signaling client
+                signalingClient = try CommunicationSignalingClient(
+                    communicationSkypeTokenProvider: tokenProvider,
+                    logger: self.options.logger
+                )
+            } catch {
+                self.options.logger.error("Signaling client initialization failed.")
+            }
+        }
+        return signalingClient
     }
```

Comment:
This looks fundamentally wrong as an async pattern. Line 147 will be executed immediately before the credential token closure completes. As an async function this needs a completion handler.